Capturing the old view transition snapshot includes descendant view-transition.

mattwoodrow opened the pull request:
#### b560a1bd3facf431c730f0c03dcffcd8923d94c8
<pre>
Capturing the old view transition snapshot includes descendant view-transition.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270056">https://bugs.webkit.org/show_bug.cgi?id=270056</a>
&lt;<a href="https://rdar.apple.com/123606877">rdar://123606877</a>&gt;

Reviewed by Tim Nguyen.

The &apos;capture the image&apos; algorithm should skip over descendant elements that
also have a view transition.

This changes snapshotNodeVisualOverflowClippedToViewport to call &apos;paint&apos; on
the RenderLayer of the element with a view-transition directly rather than
calling it on the root elements Layer (with a subtree painting root set).

It adds a new PaintFlag &apos;PaintingSkipDescendantViewTransition&apos; which skips descending
into child layers that are being separately captured in a view transition.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::snapshotNodeVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: <a href="https://commits.webkit.org/275312@main">https://commits.webkit.org/275312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b137b24cc9b09da256ed6e4763119c3b0cdb3f4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41484 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44051 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37576 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43791 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23604 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17828 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34299 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42058 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35733 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14959 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36736 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45411 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37672 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40804 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16299 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13376 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39211 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17918 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9299 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17973 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -623,6 +623,7 @@ class RenderLayer : public CanMakeSingleThreadWeakPtr<RenderLayer>, public CanMa
         PaintingChildClippingMaskPhase        = 1 << 13,
         PaintingSVGClippingMask               = 1 << 14,
         CollectingEventRegion                 = 1 << 15,
+        PaintingSkipViewTransition            = 1 << 16,
```
```suggestion
        PaintingSkipDescendantViewTransition            = 1 << 16,
```

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -262,32 +263,36 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
 
-    ASSERT(node.renderer()->hasLayer());
-    CheckedPtr layerRenderer = downcast<RenderLayerModelObject>(node.renderer());
+    auto hostWindow = (frame.document()->view() && frame.document()->view()->root()) ? frame.document()->view()->root()->hostWindow() : nullptr;
```
Should we null check or assert the page & document ?

mattwoodrow committed:
Capturing the old view transition snapshot includes descendant view-transition.
https://bugs.webkit.org/show_bug.cgi?id=270056
<rdar://123606877>

Reviewed by Tim Nguyen.

The 'capture the image' algorithm should skip over descendant elements that
also have a view transition.

This changes snapshotNodeVisualOverflowClippedToViewport to call 'paint' on
the RenderLayer of the element with a view-transition directly rather than
calling it on the root elements Layer (with a subtree painting root set).

It adds a new PaintFlag 'PaintingSkipDescendantViewTransition' which skips descending
into child layers that are being separately captured in a view transition.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::snapshotNodeVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: https://commits.webkit.org/275312@main
